### PR TITLE
Login page

### PR DESCRIPTION
**Pull request by khalosmoscato**
## Description

Created Login / Sign Up Page and sub-components needed for it.

- **Related issue:** [Insert issue number or link here]

## Type of Change

Please delete options that are not relevant:

- ✨ New feature, style (non-breaking change that adds functionality)
- ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- 📄 Documentation changes (if none of the above apply)
- 🔧 Refactor, perf, test, build, fix
- 🧹 Chore, CI, Revert

## Checklist

- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas (if necessary)
- [ ] I have updated the documentation (if necessary)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

## Screenshots (if applicable)

Screenshot 2024-10-21 at 16 53 53

### CODE REVIEW

**Comment by william-man** [comment]
Does the button being left aligned make the login form a bit inconsistent?

File: src/components/Authentication/BaseForm.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  fields: {
+    label: string;
+    type: string;
+    id: string;
+    placeholder: string;
+    required?: boolean;
+  }[];
+  buttonLabel: string;
+  onSubmit: (e: React.FormEvent) => void;
+}
```
Comment (william-man) [inline src/components/Authentication/BaseForm.tsx]: Can this be moved to the library you've created?

File: src/components/Authentication/BaseForm.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          label={field.label}
+          type={field.type}
+          id={field.id}
+          placeholder={field.placeholder}
+          required={field.required}
+        />
+      ))}
+
+      <div className="mb-4">
+      <Button label={buttonLabel} onClick={() => onSubmit} buttonType="cardButton" />
```
Comment (william-man) [inline src/components/Authentication/BaseForm.tsx]: Does this need an aria label to tell screen readers what it is?

File: src/components/Authentication/FormInput.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import React from "react";
+
+interface FormInputProps {
+  label: string;
+  type: string;
+  id: string;
+  placeholder: string;
+  required?: boolean;
+}
```
Comment (william-man) [inline src/components/Authentication/FormInput.tsx]: Does this need to be moved into the types library?

File: src/pages/Authentication.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    <Layout>
+      <div className="min-h-screen flex items-center justify-center">
+        <FormContainer>
+          {isLogin ? <LoginForm /> : <SignUpForm />}
+          <div className="text-center mt-4">
+            <p className="text-font">
+              {isLogin ? "Don't have an account?" : "Already have an account?"}{" "}
+              <button
+                onClick={toggleForm}
+                className="text-zubiHighlight font-bold underline ml-1"
```
Comment (william-man) [inline src/pages/Authentication.tsx:24]: Does this need an aria label?

**Comment by khalosmoscato** [comment]
image
@william-man centred the button and also added aria label for accessibility. 
Moved interface to our library.
@jackcasstlesjones made the `Log In` link stand out in the menu by adding a yellow button-like background, which means I also had to change the hoover effect otherwise the text would become yellow on yellow, but not it becomes white.

**Comment by khalosmoscato** [comment]
@jackcasstlesjones @william-man can one of you help me with best practice of sanitise and validate before we merge this to main. Should we use an external libraries, or implement our own validation functions for each input field?

**Comment by khalosmoscato** [comment]
Error message when incorrect login credentials are entered:
Screenshot 2024-10-23 at 14 43 10

**Comment by khalosmoscato** [comment]
When login is successful, redirects to home page and `My Account` becomes `Student Dashboard` or `Tutor Dashboard`.

This is a bit more future proof, because when we will have the tutors login in the future, they will have access to a different dashboard then students. 

I created an interface that defines the user role:

`export interface User {
  name?: string; // Optional in case no name is provided
  email: string;
  role: 'student' | 'tutor'; 
}`

Screenshot 2024-10-23 at 14 43 37

**Comment by khalosmoscato** [comment]
I need to: improve the logout functionality to not redirect the user back to the login page, but refresh the homepage instead.

**Review by william-man — APPROVED** [review]
lgtm

**Pull request merged**